// @pkmn/sim/build/sim/dex-species.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Species = void 0;
const dex_data_1 = require("./dex-data");
class Species extends dex_data_1.BasicEffect {
    constructor(data, ...moreData) {
        super(data, ...moreData);
        data = this;
        this.fullname = `pokemon: ${data.name}`;
        this.effectType = 'Pokemon';
        this.id = data.id;
        this.name = data.name;
        this.baseSpecies = data.baseSpecies || this.name;
        this.forme = data.forme || '';
        this.baseForme = data.baseForme || '';
        this.cosmeticFormes = data.cosmeticFormes || undefined;
        this.otherFormes = data.otherFormes || undefined;
        this.formeOrder = data.formeOrder || undefined;
        this.spriteid = data.spriteid ||
            (dex_data_1.toID(this.baseSpecies) + (this.baseSpecies !== this.name ? `-${dex_data_1.toID(this.forme)}` : ''));
        this.abilities = data.abilities || { 0: "" };
        this.types = data.types || ['???'];
        this.addedType = data.addedType || undefined;
        this.prevo = data.prevo || '';
        this.tier = data.tier || '';
        this.doublesTier = data.doublesTier || '';
        this.evos = data.evos || [];
        this.evoType = data.evoType || undefined;
        this.evoMove = data.evoMove || undefined;
        this.evoLevel = data.evoLevel || undefined;
        this.nfe = data.nfe || false;
        this.eggGroups = data.eggGroups || [];
        this.canHatch = data.canHatch || false;
        this.gender = data.gender || '';
        this.genderRatio = data.genderRatio || (this.gender === 'M' ? { M: 1, F: 0 } :
            this.gender === 'F' ? { M: 0, F: 1 } :
                this.gender === 'N' ? { M: 0, F: 0 } :
                    { M: 0.5, F: 0.5 });
        this.requiredItem = data.requiredItem || undefined;
        this.requiredItems = this.requiredItems || (this.requiredItem ? [this.requiredItem] : undefined);
        this.baseStats = data.baseStats || { hp: 0, atk: 0, def: 0, spa: 0, spd: 0, spe: 0 };
        this.bst = this.baseStats.hp + this.baseStats.atk + this.baseStats.def +
            this.baseStats.spa + this.baseStats.spd + this.baseStats.spe;
        this.weightkg = data.weightkg || 0;
        this.weighthg = this.weightkg * 10;
        this.heightm = data.heightm || 0;
        this.color = data.color || '';
        this.unreleasedHidden = data.unreleasedHidden || false;
        this.maleOnlyHidden = !!data.maleOnlyHidden;
        this.maxHP = data.maxHP || undefined;
        this.isMega = !!(this.forme && ['Mega', 'Mega-X', 'Mega-Y'].includes(this.forme)) || undefined;
        this.canGigantamax = data.canGigantamax || undefined;
        this.gmaxUnreleased = !!data.gmaxUnreleased;
        this.cannotDynamax = !!data.cannotDynamax;
        this.battleOnly = data.battleOnly || (this.isMega ? this.baseSpecies : undefined);
        this.changesFrom = data.changesFrom ||
            (this.battleOnly !== this.baseSpecies ? this.battleOnly : this.baseSpecies);
        if (Array.isArray(data.changesFrom))
            this.changesFrom = data.changesFrom[0];
        if (!this.gen && this.num >= 1) {
            if (this.num >= 810 || ['Gmax', 'Galar', 'Galar-Zen'].includes(this.forme)) {
                this.gen = 8;
            }
            else if (this.num >= 722 || this.forme.startsWith('Alola') || this.forme === 'Starter') {
                this.gen = 7;
            }
            else if (this.forme === 'Primal') {
                this.gen = 6;
                this.isPrimal = true;
                this.battleOnly = this.baseSpecies;
            }
            else if (this.num >= 650 || this.isMega) {
                this.gen = 6;
            }
            else if (this.num >= 494) {
                this.gen = 5;
            }
            else if (this.num >= 387) {
                this.gen = 4;
            }
            else if (this.num >= 252) {
                this.gen = 3;
            }
            else if (this.num >= 152) {
                this.gen = 2;
            }
            else {
                this.gen = 1;
            }
        }
    }
}
exports.Species = Species;
