// @pkmn/sim/build/sim/dex-items.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Item = void 0;
const dex_data_1 = require("./dex-data");
class Item extends dex_data_1.BasicEffect {
    constructor(data, ...moreData) {
        super(data, ...moreData);
        data = this;
        this.fullname = `item: ${this.name}`;
        this.effectType = 'Item';
        this.fling = data.fling || undefined;
        this.onDrive = data.onDrive || undefined;
        this.onMemory = data.onMemory || undefined;
        this.megaStone = data.megaStone || undefined;
        this.megaEvolves = data.megaEvolves || undefined;
        this.zMove = data.zMove || undefined;
        this.zMoveType = data.zMoveType || undefined;
        this.zMoveFrom = data.zMoveFrom || undefined;
        this.itemUser = data.itemUser || undefined;
        this.isBerry = !!data.isBerry;
        this.ignoreKlutz = !!data.ignoreKlutz;
        this.onPlate = data.onPlate || undefined;
        this.isGem = !!data.isGem;
        this.isPokeball = !!data.isPokeball;
        if (!this.gen) {
            if (this.num >= 689) {
                this.gen = 7;
            }
            else if (this.num >= 577) {
                this.gen = 6;
            }
            else if (this.num >= 537) {
                this.gen = 5;
            }
            else if (this.num >= 377) {
                this.gen = 4;
            }
            else {
                this.gen = 3;
            }
            // Due to difference in gen 2 item numbering, gen 2 items must be
            // specified manually
        }
        if (this.isBerry)
            this.fling = { basePower: 10 };
        if (this.id.endsWith('plate'))
            this.fling = { basePower: 90 };
        if (this.onDrive)
            this.fling = { basePower: 70 };
        if (this.megaStone)
            this.fling = { basePower: 80 };
        if (this.onMemory)
            this.fling = { basePower: 50 };
    }
}
exports.Item = Item;
